command-line/lib/Command/Key/WalletId.hs

```diff
@@ -15,62 +16,98 @@ import Prelude hiding
 
 data Cmd = WalletId
-    deriving (Show)
+    { spending :: Maybe (Script Cosigner)
+    , staking :: Maybe (Script Cosigner)
+    } deriving (Show)
 
 mod :: (Cmd -> parent) -> Mod CommandFields parent
 mod liftCmd = command "walletid" $
     info (helper <*> fmap liftCmd parser) $ mempty
         <> progDesc "Shows the cardano-wallet wallet ID for a given key"
         <> footerDoc (Just $ string $ mconcat
             [ "A wallet ID is a 40-digit hexadecimal string derived "
-            , "from the wallet’s mnemonic. It is used by the cardano-wallet "
+            , "from the wallet’s key. It is used by the cardano-wallet "
             , "server to refer to specific wallets.\n\n"
-            , "This key can be either an extended root key "
+            , "For shelley wallets the key can be either an extended root key "
             , "(full multi-account wallets), or an extended account key "
             , "(single-account wallets).\n\n"
-            , "Private or public extended keys are accepted -- they will have "
-            , "the same wallet ID.\n\n"
-            , "The bech32-encoded key is read from standard input.\n"
+            , "In the latter case either private or public extended key is accepted"
+            , "-- they will have the same wallet ID.\n\n"
+            , "The bech32-encoded key is read from standard input.\n\n"
+            , "In case of a shared-wallet the wallet id is calculated based on"
+            , "both private (public) extended account key, payment template script and"
+            , "staking template script. Each signature in any template script is denoted"
+            , "by cosigner#number.\n"
             ])
   where
-    parser = pure WalletId
+    parser = WalletId
+        <$> optional scriptTemplateSpendingArg
+        <*> optional scriptTemplateStakingArg
 
 run :: Cmd -> IO ()
-run WalletId = do
+run WalletId{spending,staking} = do
     (hrp, bytes) <- hGetBech32 stdin allowedPrefixes
     guardBytes hrp bytes
-    hPutBytes stdout (hashWalletId $ payloadToHash hrp bytes) EBase16
+    let bs = payloadToHash hrp bytes
+    when ( hrp `elem` [CIP5.acct_shared_xvk, CIP5.acct_shared_xsk] &&
+           isNothing spending ) $
+        fail "shared wallet needs to have at least spending script specified"
+    let walletid =
+            if hrp `elem` [CIP5.acct_shared_xvk, CIP5.acct_shared_xsk] then
+                hashWalletId $
+                bs <>
+                scriptTemplateToBs (fromJust spending) <>
+                maybe mempty scriptTemplateToBs staking
+            else
+                hashWalletId bs
+    hPutBytes stdout walletid EBase16
   where
     allowedPrefixes =
         [ CIP5.root_xsk
         , CIP5.root_xvk
-        , CIP5.root_shared_xsk
-        , CIP5.root_shared_xvk
         , CIP5.acct_xvk
         , CIP5.acct_xsk
         , CIP5.acct_shared_xvk
         , CIP5.acct_shared_xsk
         ]
 
+    scriptTemplateToBs = T.encodeUtf8 . T.pack . show
```

Comment:
Hm, using `show` here will probably make it very difficult for non-Haskell code to generate matching wallet IDs 🤔. Then again, wallet IDs are an internal affair, so we don't have to cater to that.

Still, I was wondering whether we should use `serializeScript` to get a canonical serialization before hashing, but this function expects a `Script KeyHash`, not `Script Cosigner`.

Proposal: How about renaming `scriptTemplateToBs` to
```hs
serializeScriptTemplate :: Script Cosigner -> ByteString
```
and exporting it on the top level from `Cardano.Address.Script`? In this way, the function that serializes a script template is at least "fully specified and exported" (and we can later reuse it in cardano-wallet).

Reply:
good idea.

1. now we will calculate byte payload by utf8 encoding of strings that are used in json/command line. so
```
cosigner#0
all [cosigner#1, cossigner#0, active_from 1000]
```
rather than show instance. It will be easier to reconstruct it for any user doing this outside cardano-addresses

2. I have exposed `sharedWalletId` to be used in cardano-wallet to promote better reuse of cardano-addresses and avoid unneeded redundancy